{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create SRA submission sheet\n",
    "This Python Jupyter notebook creates a BioSample submission sheet for the Sequence Read Archive:\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import glob\n",
    "import itertools\n",
    "import os\n",
    "\n",
    "import natsort\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display full columns of data frames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the configuration for the uploads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('upload_config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the Illumina runs to submit:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Illumina runs from barcode_runs_Crowe_antibodies.csv\n",
      "Here are the first few entries in the Illumina runs:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>date</th>\n",
       "      <th>experiment</th>\n",
       "      <th>antibody</th>\n",
       "      <th>concentration</th>\n",
       "      <th>concentration_units</th>\n",
       "      <th>group</th>\n",
       "      <th>selection</th>\n",
       "      <th>frac_escape</th>\n",
       "      <th>cells_sorted</th>\n",
       "      <th>R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib1</td>\n",
       "      <td>expt_13-16-none-0-reference</td>\n",
       "      <td>200720</td>\n",
       "      <td>expt_13-16</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ng_per_mL</td>\n",
       "      <td>none</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160000000.0</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/*R1*.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib2</td>\n",
       "      <td>expt_13-16-none-0-reference</td>\n",
       "      <td>200720</td>\n",
       "      <td>expt_13-16</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ng_per_mL</td>\n",
       "      <td>none</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160000000.0</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib2_ref/*R1*.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib1</td>\n",
       "      <td>expt_7-11-none-0-reference</td>\n",
       "      <td>200616</td>\n",
       "      <td>expt_7-11</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ng_per_mL</td>\n",
       "      <td>none</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160000000.0</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200618_D00300_0980_AHFCMJBCX3/Unaligned/Project_agreaney/Sample_expt_7-11_lib1_ACE2pos/*.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib2</td>\n",
       "      <td>expt_7-11-none-0-reference</td>\n",
       "      <td>200616</td>\n",
       "      <td>expt_7-11</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ng_per_mL</td>\n",
       "      <td>none</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160000000.0</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200618_D00300_0980_AHFCMJBCX3/Unaligned/Project_agreaney/Sample_expt_7-11_lib2_ACE2pos/*.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib1</td>\n",
       "      <td>expt_22-23-none-0-reference</td>\n",
       "      <td>200810</td>\n",
       "      <td>expt_22-23</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ng_per_mL</td>\n",
       "      <td>none</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160000000.0</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200812_D00300_1026_BHFFJTBCX3/Unaligned/Project_agreaney/Sample_expt_22-23_lib1_ref/*R1*.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library                       sample    date  experiment antibody  \\\n",
       "0    lib1  expt_13-16-none-0-reference  200720  expt_13-16     none   \n",
       "1    lib2  expt_13-16-none-0-reference  200720  expt_13-16     none   \n",
       "2    lib1   expt_7-11-none-0-reference  200616   expt_7-11     none   \n",
       "3    lib2   expt_7-11-none-0-reference  200616   expt_7-11     none   \n",
       "4    lib1  expt_22-23-none-0-reference  200810  expt_22-23     none   \n",
       "\n",
       "   concentration concentration_units group  selection  frac_escape  \\\n",
       "0              0           ng_per_mL  none  reference          NaN   \n",
       "1              0           ng_per_mL  none  reference          NaN   \n",
       "2              0           ng_per_mL  none  reference          NaN   \n",
       "3              0           ng_per_mL  none  reference          NaN   \n",
       "4              0           ng_per_mL  none  reference          NaN   \n",
       "\n",
       "   cells_sorted  \\\n",
       "0   160000000.0   \n",
       "1   160000000.0   \n",
       "2   160000000.0   \n",
       "3   160000000.0   \n",
       "4   160000000.0   \n",
       "\n",
       "                                                                                                                                R1  \n",
       "0  /shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/*R1*.fastq.gz  \n",
       "1  /shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib2_ref/*R1*.fastq.gz  \n",
       "2  /shared/ngs/illumina/agreaney/200618_D00300_0980_AHFCMJBCX3/Unaligned/Project_agreaney/Sample_expt_7-11_lib1_ACE2pos/*.fastq.gz  \n",
       "3  /shared/ngs/illumina/agreaney/200618_D00300_0980_AHFCMJBCX3/Unaligned/Project_agreaney/Sample_expt_7-11_lib2_ACE2pos/*.fastq.gz  \n",
       "4  /shared/ngs/illumina/agreaney/200812_D00300_1026_BHFFJTBCX3/Unaligned/Project_agreaney/Sample_expt_22-23_lib1_ref/*R1*.fastq.gz  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Reading Illumina runs from {config['barcode_runs']}\")\n",
    "\n",
    "illumina_runs = pd.read_csv(config['barcode_runs'])\n",
    "\n",
    "print('Here are the first few entries in the Illumina runs:')\n",
    "illumina_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for the barcode runs, initially in \"tidy\" format with one FASTQ file per row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in tidy format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-16-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L001_R1_002.fastq.gz</td>\n",
       "      <td>expt_13-16_lib1_ref_AAGGTACA_L001_R1_002.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-16-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L002_R1_004.fastq.gz</td>\n",
       "      <td>expt_13-16_lib1_ref_AAGGTACA_L002_R1_004.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-16-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L002_R1_002.fastq.gz</td>\n",
       "      <td>expt_13-16_lib1_ref_AAGGTACA_L002_R1_002.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-16-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L001_R1_005.fastq.gz</td>\n",
       "      <td>expt_13-16_lib1_ref_AAGGTACA_L001_R1_005.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-16-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L002_R1_005.fastq.gz</td>\n",
       "      <td>expt_13-16_lib1_ref_AAGGTACA_L002_R1_005.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession                        library_ID  \\\n",
       "0        SAMN16054076  lib1_expt_13-16-none-0-reference   \n",
       "1        SAMN16054076  lib1_expt_13-16-none-0-reference   \n",
       "2        SAMN16054076  lib1_expt_13-16-none-0-reference   \n",
       "3        SAMN16054076  lib1_expt_13-16-none-0-reference   \n",
       "4        SAMN16054076  lib1_expt_13-16-none-0-reference   \n",
       "\n",
       "                                                                                         title  \\\n",
       "0  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference   \n",
       "1  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference   \n",
       "2  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference   \n",
       "3  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference   \n",
       "4  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference   \n",
       "\n",
       "  library_strategy library_source library_selection library_layout  platform  \\\n",
       "0         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "      instrument_model                 design_description filetype  \\\n",
       "0  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "1  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "2  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "3  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "4  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                                                                                                                                     filename_fullpath  \\\n",
       "0  /shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L001_R1_002.fastq.gz   \n",
       "1  /shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L002_R1_004.fastq.gz   \n",
       "2  /shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L002_R1_002.fastq.gz   \n",
       "3  /shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L001_R1_005.fastq.gz   \n",
       "4  /shared/ngs/illumina/agreaney/200727_D00300_1012_AHFFMGBCX3/Unaligned/Project_agreaney/Sample_expt_13-16_lib1_ref/expt_13-16_lib1_ref_AAGGTACA_L002_R1_005.fastq.gz   \n",
       "\n",
       "                                            filename  \n",
       "0  expt_13-16_lib1_ref_AAGGTACA_L001_R1_002.fastq.gz  \n",
       "1  expt_13-16_lib1_ref_AAGGTACA_L002_R1_004.fastq.gz  \n",
       "2  expt_13-16_lib1_ref_AAGGTACA_L002_R1_002.fastq.gz  \n",
       "3  expt_13-16_lib1_ref_AAGGTACA_L001_R1_005.fastq.gz  \n",
       "4  expt_13-16_lib1_ref_AAGGTACA_L002_R1_005.fastq.gz  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_tidy = (\n",
    "    illumina_runs\n",
    "    .assign(\n",
    "        biosample_accession=lambda x: config['biosample_accession'],\n",
    "        library_ID=lambda x: x['library'] + '_' + x['sample'],\n",
    "        title=lambda x: 'SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for ' + x['sample'],\n",
    "        library_strategy='AMPLICON',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='PCR',\n",
    "        library_layout='single',\n",
    "        platform='ILLUMINA',\n",
    "        instrument_model='Illumina HiSeq 2500',\n",
    "        design_description='PCR of barcodes from RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['R1'].str.split(';')\n",
    "                                    .map(lambda flist: list(itertools.chain.from_iterable(glob.glob(f) for f in flist))),       \n",
    "        )\n",
    "    .explode('filename_fullpath')\n",
    "    .assign(filename_fullpath=lambda x: x['filename_fullpath'].str.strip(),\n",
    "            filename=lambda x: x['filename_fullpath'].map(os.path.basename))\n",
    "    .drop(columns=illumina_runs.columns)\n",
    "    .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "assert submissions_tidy['filename_fullpath'].map(os.path.isfile).all()\n",
    "\n",
    "print('Here are the first few submission entries in tidy format:')\n",
    "submissions_tidy.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the actual submission, we need a \"wide\" data frame that for each unique `sample_name` / `library_ID` gives all of the files each in different columns.\n",
    "These should be files without the full path.\n",
    "\n",
    "First, look at how many files there are for each sample / library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>n_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_22-23-none-0-reference</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_22-23-none-0-reference</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_3-6-none-0-reference</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_3-6-none-0-reference</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_7-11-none-0-reference</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_7-11-none-0-reference</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-16-none-0-reference</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_13-16-none-0-reference</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_15-COV2-2499-400-escape</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_23-COV2-2096-400-escape</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_13-COV2-2050-400-escape</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_23-COV2-2096-400-escape</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_15-COV2-2499-400-escape</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_5-COV2-2479-400-escape</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_14-COV2-2094-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_3-CR3022-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_16-COV2-2677-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_10-COV2-2832-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_10-COV2-2832-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_11-COV2-2082-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_11-COV2-2082-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_9-COV2-2165-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_5-COV2-2479-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_3-CR3022-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_16-COV2-2677-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_14-COV2-2094-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-COV2-2050-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_9-COV2-2165-400-escape</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_accession                         library_ID  n_files\n",
       "0         SAMN16054076   lib1_expt_22-23-none-0-reference       34\n",
       "1         SAMN16054076   lib2_expt_22-23-none-0-reference       32\n",
       "2         SAMN16054076     lib1_expt_3-6-none-0-reference       16\n",
       "3         SAMN16054076     lib2_expt_3-6-none-0-reference       16\n",
       "4         SAMN16054076    lib2_expt_7-11-none-0-reference       16\n",
       "5         SAMN16054076    lib1_expt_7-11-none-0-reference       16\n",
       "6         SAMN16054076   lib1_expt_13-16-none-0-reference       12\n",
       "7         SAMN16054076   lib2_expt_13-16-none-0-reference       10\n",
       "8         SAMN16054076  lib2_expt_15-COV2-2499-400-escape        4\n",
       "9         SAMN16054076  lib2_expt_23-COV2-2096-400-escape        4\n",
       "10        SAMN16054076  lib2_expt_13-COV2-2050-400-escape        4\n",
       "11        SAMN16054076  lib1_expt_23-COV2-2096-400-escape        4\n",
       "12        SAMN16054076  lib1_expt_15-COV2-2499-400-escape        4\n",
       "13        SAMN16054076   lib2_expt_5-COV2-2479-400-escape        3\n",
       "14        SAMN16054076  lib2_expt_14-COV2-2094-400-escape        2\n",
       "15        SAMN16054076      lib2_expt_3-CR3022-400-escape        2\n",
       "16        SAMN16054076  lib2_expt_16-COV2-2677-400-escape        2\n",
       "17        SAMN16054076  lib1_expt_10-COV2-2832-400-escape        2\n",
       "18        SAMN16054076  lib2_expt_10-COV2-2832-400-escape        2\n",
       "19        SAMN16054076  lib2_expt_11-COV2-2082-400-escape        2\n",
       "20        SAMN16054076  lib1_expt_11-COV2-2082-400-escape        2\n",
       "21        SAMN16054076   lib1_expt_9-COV2-2165-400-escape        2\n",
       "22        SAMN16054076   lib1_expt_5-COV2-2479-400-escape        2\n",
       "23        SAMN16054076      lib1_expt_3-CR3022-400-escape        2\n",
       "24        SAMN16054076  lib1_expt_16-COV2-2677-400-escape        2\n",
       "25        SAMN16054076  lib1_expt_14-COV2-2094-400-escape        2\n",
       "26        SAMN16054076  lib1_expt_13-COV2-2050-400-escape        2\n",
       "27        SAMN16054076   lib2_expt_9-COV2-2165-400-escape        2"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(submissions_tidy\n",
    " .groupby(['biosample_accession', 'library_ID'])\n",
    " .aggregate(n_files=pd.NamedAgg('filename_fullpath', 'count'))\n",
    " .sort_values('n_files', ascending=False)\n",
    " .reset_index()\n",
    " )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now make the wide submission data frame.\n",
    "Note we keep only the filename column with the path lacking the full directory information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in wide format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>...</th>\n",
       "      <th>filename25</th>\n",
       "      <th>filename26</th>\n",
       "      <th>filename27</th>\n",
       "      <th>filename28</th>\n",
       "      <th>filename29</th>\n",
       "      <th>filename30</th>\n",
       "      <th>filename31</th>\n",
       "      <th>filename32</th>\n",
       "      <th>filename33</th>\n",
       "      <th>filename34</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_13-16-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_13-16-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_7-11-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_7-11-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib2_expt_7-11-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_7-11-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN16054076</td>\n",
       "      <td>lib1_expt_22-23-none-0-reference</td>\n",
       "      <td>SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_22-23-none-0-reference</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>...</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L001_R1_014.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L002_R1_009.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L001_R1_001.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L002_R1_001.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L002_R1_011.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L001_R1_004.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L002_R1_013.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L002_R1_017.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L002_R1_002.fastq.gz</td>\n",
       "      <td>expt_22-23_lib1_ref_ACCACTGT_L002_R1_015.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession                        library_ID  \\\n",
       "0        SAMN16054076  lib1_expt_13-16-none-0-reference   \n",
       "1        SAMN16054076  lib2_expt_13-16-none-0-reference   \n",
       "2        SAMN16054076   lib1_expt_7-11-none-0-reference   \n",
       "3        SAMN16054076   lib2_expt_7-11-none-0-reference   \n",
       "4        SAMN16054076  lib1_expt_22-23-none-0-reference   \n",
       "\n",
       "                                                                                         title  \\\n",
       "0  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference   \n",
       "1  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_13-16-none-0-reference   \n",
       "2   SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_7-11-none-0-reference   \n",
       "3   SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_7-11-none-0-reference   \n",
       "4  SARS-CoV-2 RBD mutational antig Illumina barcode sequencing for expt_22-23-none-0-reference   \n",
       "\n",
       "  library_strategy library_source library_selection library_layout  platform  \\\n",
       "0         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "      instrument_model                 design_description  ...  \\\n",
       "0  Illumina HiSeq 2500  PCR of barcodes from RBD variants  ...   \n",
       "1  Illumina HiSeq 2500  PCR of barcodes from RBD variants  ...   \n",
       "2  Illumina HiSeq 2500  PCR of barcodes from RBD variants  ...   \n",
       "3  Illumina HiSeq 2500  PCR of barcodes from RBD variants  ...   \n",
       "4  Illumina HiSeq 2500  PCR of barcodes from RBD variants  ...   \n",
       "\n",
       "                                          filename25  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L001_R1_014.fastq.gz   \n",
       "\n",
       "                                          filename26  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L002_R1_009.fastq.gz   \n",
       "\n",
       "                                          filename27  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                          filename28  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L002_R1_001.fastq.gz   \n",
       "\n",
       "                                          filename29  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L002_R1_011.fastq.gz   \n",
       "\n",
       "                                          filename30  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L001_R1_004.fastq.gz   \n",
       "\n",
       "                                          filename31  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L002_R1_013.fastq.gz   \n",
       "\n",
       "                                          filename32  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L002_R1_017.fastq.gz   \n",
       "\n",
       "                                          filename33  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L002_R1_002.fastq.gz   \n",
       "\n",
       "                                          filename34  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3                                                NaN  \n",
       "4  expt_22-23_lib1_ref_ACCACTGT_L002_R1_015.fastq.gz  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .assign(\n",
    "        filename_count=lambda x: x.groupby(['biosample_accession', 'library_ID'])['filename'].cumcount() + 1,\n",
    "        filename_col=lambda x: 'filename' + x['filename_count'].map(lambda c: str(c) if c > 1 else '')\n",
    "        )\n",
    "    .pivot(\n",
    "        index='library_ID',\n",
    "        columns='filename_col',\n",
    "        values='filename',\n",
    "        )\n",
    "    )\n",
    "\n",
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .drop(columns=['filename_fullpath', 'filename'])\n",
    "    .drop_duplicates()\n",
    "    .merge(submissions_wide[natsort.natsorted(submissions_wide.columns)],\n",
    "           on='library_ID',\n",
    "           validate='one_to_one',\n",
    "           )\n",
    "    )\n",
    "\n",
    "print('Here are the first few submission entries in wide format:')\n",
    "submissions_wide.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now write the wide submissions data frame to a `*.tsv` file that can be used for uploading the SRA submission website:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "submissions_spreadsheet = 'SRA_submission_spreadsheet.tsv'\n",
    "\n",
    "submissions_wide.to_csv(submissions_spreadsheet, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also want to write a file with all of the FASTQ files that are in the submission spreadsheet:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing list of all the FASTQ files to upload to FASTQs_to_upload.csv\n"
     ]
    }
   ],
   "source": [
    "fastq_file_list = 'FASTQs_to_upload.csv'\n",
    "\n",
    "print(f\"Writing list of all the FASTQ files to upload to {fastq_file_list}\")\n",
    "\n",
    "(submissions_tidy\n",
    " [['filename_fullpath', 'filename']]\n",
    " .to_csv(fastq_file_list, index=False)\n",
    " )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
